### ISSUE TITLE
Signal connections established with `+=` are not automatically removed when receiver is destroyed

### ISSUE THREAD

**Issue by derkork** [godotengine/godot#70414]
### Godot version

v4.0.beta9.mono.official [e780dc332]

### System information

Windows 10

### Issue description

Assume you create a custom node class which emits a signal every second:

```csharp
// Sender.cs
	[Signal]
	public delegate void MySignalEventHandler();
	
	public override void _Ready()
	{
		GetTree().CreateTimer(1).Timeout += OnTimeout;
	}

	
	private void OnTimeout()
	{
		EmitSignal(nameof(MySignal));
		GetTree().CreateTimer(1).Timeout += OnTimeout;
	}
```

Now you have a receiver which connects to this signal either with the classic "Connect" method or with the new `+=` syntax. The receiver destroys itself after a few seconds:

```csharp
// Receiver.cs
 public override void _Ready()
    {
        if (UseConnect)
        {
            Sender.Connect(nameof(Sender.MySignal), Callable.From(Receive));
        }
        else
        {
            Sender.MySignal += Receive;
        }
        
        // destroy after the specified time
        if (DestroyAfterSeconds > 0)
        {
            GetTree().CreateTimer(DestroyAfterSeconds).Timeout += QueueFree;
        }
    }
```

Either way, I would expect that the signal is automatically disconnected when when the receiver is destroyed. This works when using the `Connect` method, but doesn't work when using the `+=` method. In the latter case the signal connection stays and the sender throws an exception when it emits the signal the next time:

```
E 0:00:10:0018   void Godot.Bridge.ScriptManagerBridge.RaiseEventSignal(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_bool* ): System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'sasake.Receiver'.
  <C++ Error>    Exception
  <C++ Source>   /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs:73 @ IntPtr Godot.Object.GetPtr(Godot.Object )()
  <Stack Trace>  Object.base.cs:73 @ IntPtr Godot.Object.GetPtr(Godot.Object )()
                 Node.cs:590 @ Godot.StringName Godot.Node.GetName()()
                 Node.cs:290 @ Godot.StringName Godot.Node.get_Name()()
                 Receiver.cs:31 @ void sasake.Receiver.Receive()()
                 Sender_ScriptSignals.generated.cs:26 @ void Sender.RaiseGodotClassSignalCallbacks(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs )()
                 ScriptManagerBridge.cs:341 @ void Godot.Bridge.ScriptManagerBridge.RaiseEventSignal(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_bool* )()

```

It looks like the sender tries to invoke a callback on the now disposed receiver. Again this works fine when using `Connect` to connect the signal.

### Steps to reproduce

In the example project open the sender.tscn scene and run it. You can see in the output that the receiver receives two signals before it self destructs. 

![image](https://user-images.githubusercontent.com/327257/208986911-ffdc1ba9-8fb2-43a6-a0ab-65f4d418351e.png)

After that you can see the exception raised in the sender:

![image](https://user-images.githubusercontent.com/327257/208987025-5ad4c9e5-0c27-4777-b165-f64f63bd654c.png)

When you switch the receiver to use "Connect" using the inspector: 

![image](https://user-images.githubusercontent.com/327257/208987205-dd78edd4-d205-4b09-b34e-56ba63236510.png)

Then everything works as expected.

### Minimal reproduction project

[example.zip](https://github.com/godotengine/godot/files/10280604/example.zip)

**Comment by Calinou** [comment]
Is this operator overload for connecting signals documented anywhere? I don't remember seeing this in GDScript. Why was it added?

**Comment by derkork** [comment]
> Is this operator overload for connecting signals documented anywhere? I don't remember seeing this in GDScript. Why was it added?

At least in the [getting started section](https://docs.godotengine.org/en/latest/getting_started/step_by_step/signals.html#complete-script) it is done with `+=`, e.g. `timer.Timeout += OnTimerTimeout;`.

**Comment by derkork** [comment]
It's probably caused by the generated implementation for the signal:

```csharp
    private global::Sender.MySignalEventHandler backing_MySignal;
    /// <inheritdoc cref="global::Sender.MySignalEventHandler"/>
    public event global::Sender.MySignalEventHandler MySignal {
        add => backing_MySignal += value;
        remove => backing_MySignal -= value;
}
```

This wil just add the new event handler as a multicast target to the backing event handler, so the engine will not even know that the signal is connected. I checked that by printing the amount of signal connections in the sender. When using the `+=` method, it reports 1 signal being connected (probably some internal connection) while when using the connect method it reports 2 signals being connected (the internal one and the one we added with `Connect`). 

One solution could be to actually call "Connect" / Disconnect in the generated method and not use a "backing" signal delegate:

```csharp
    public event global::Sender.MySignalEventHandler MySignal {
        add => Connect(nameof(MySignal), Callable.From(value));
        remove => Disconnect(nameof(MySignal), Callable.From(value)); // this will actually not work as it is a new callable
}
```

but you would need a way to transform the delegate type into a Callable plus you would need some kind of bookkeeping between the `Callable` and the original delegate that was used to create the callable so you can do a proper disconnect.

**Comment by markdibarry** [comment]
> Is this operator overload for connecting signals documented anywhere? I don't remember seeing this in GDScript. Why was it added?

It is the current recommended way of connecting signals in C# and was introduced to replicate the built-in observer pattern in the C# language. You can see [some examples here](https://docs.godotengine.org/en/latest/classes/class_object.html#class-object-method-connect).

**Comment by lorenzo-arena** [comment]
Are there any news about this issue?

**Comment by Ark2000** [comment]
additional: it also throws `System.ObjectDisposedException` if you are using lambda or local function in `Connect`

like this: 
```csharp
Sender.Connect(nameof(Sender.MySignal), Callable.From(()=>{GD.Print("hello")}));
```
or 
```csharp
void sayHello()
{
    GD.Print("hello")
}
Sender.Connect(nameof(Sender.MySignal), Callable.From(sayHello));
```

**Comment by fabianPas** [comment]
Any news on this problem? I'm experiencing it as well and think it's quite a nasty one as the documentation is filled with event connecting examples instead of .Connect().

**Comment by AThousandShips** [comment]
See the linked PR:
* #73730

**Comment by fabianPas** [comment]
I see, thank you very much!

**Comment by hunterloftis** [comment]
This issue is over a year old. Given the nasty set of bugs it silently produces, should it be either:

1. Prioritized for a fix or
2. Prioritized for correct documentation (replacing all examples of `+=` with a safer mechanism?)

At the very least, it would be useful to provide a non-`+=` alternative in the "bound values" section, where that's the only mechanism documented at all:

https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_signals.html#bound-values

**Comment by 31** [comment]
About a month ago, I contributed some PRs to the C# signals doc to describe the behavior with some more detail. However, it's only available in `latest (4.3)`: https://docs.godotengine.org/en/latest/tutorials/scripting/c_sharp/c_sharp_signals.html. It includes a general example for how to use `+=` with closures in a way that keeps the delegate around to pass to `-=` in a cleanup method.

PTAL and send some followup PRs if you can make it clearer! I've gotten feedback that the example is a bit too contrived and makes it hard to follow the rest of it, but I haven't had a chance to go back and see if I can make it better by giving more context or using a more gamedev-linked example.

**Comment by hunterloftis** [comment]
@31 thanks for the link, your additions are a huge improvement to the docs in that section!

**Comment by hunterloftis** [comment]
Agreed @tm76, given that signals are such a core part of Godot, most C# users would likely be surprised that they aren't rock solid / have so many gotchas.

**Comment by HymnStudio** [comment]
> About a month ago, I contributed some PRs to the C# signals doc to describe the behavior with some more detail. However, it's only available in `latest (4.3)`: https://docs.godotengine.org/en/latest/tutorials/scripting/c_sharp/c_sharp_signals.html. It includes a general example for how to use `+=` with closures in a way that keeps the delegate around to pass to `-=` in a cleanup method.
> 
> PTAL and send some followup PRs if you can make it clearer! I've gotten feedback that the example is a bit too contrived and makes it hard to follow the rest of it, but I haven't had a chance to go back and see if I can make it better by giving more context or using a more gamedev-linked example.

As you said in the latest doc that using Connect will disconnect automatically with custom signals, while using += needs manual disconnection, I wonder whether connecting through Godot Editor Inspector equals using Connect method？Thank you！

**Comment by dante-signal31** [comment]
@HymnStudio I think editor signal connections indeed equals those using Connect().

The test I've done: 

1. I've put a QueueFree() in a receiver object so it is removed after first signal emission. 
2. I emitted once the signal and receiver has been properly removed. 
3. In the second signal emission debugger does not throw any error. So I guess signal receiver was properly disconnected from emissor when QueueFree() was called.

**Comment by lfod1997** [comment]
> additional: it also throws `System.ObjectDisposedException` when the listener object is destroyed if you are using lambda or local function in `Connect`
> 
> like this:
> 
> Sender.Connect(nameof(Sender.MySignal), Callable.From(()=>{GD.Print("hello")}));
> 
> or
> 
> void sayHello()
> {
>     GD.Print("hello")
> }
> Sender.Connect(nameof(Sender.MySignal), Callable.From(sayHello));

@Ark2000 Not true.

1. Invalid C#, have you tried your code before posting?
2. Lambda `()=>{GD.Print("hello");}` does not capture, will be invoked, won't throw. Expected behavior.
3. Instance method `void sayHello() {GD.Print("hello");}` is bound to your "listener object" and, as it's destroyed, the method won't be called. Also expected behavior.

**Comment by AfroSpartan** [comment]
Just hit this issue with subscribing to EventHandlers, I'm guessing these aren't intended to be supported?

**Comment by viksl** [comment]
@AfroSpartan All the stuff how to use signals with c# be it with connect or += and -= are nicely written in the docs: https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_signals.html#c-signals linked earlier. You can use c# events, godot signals, combine it in your code.

**Comment by AfroSpartan** [comment]
> [@AfroSpartan](https://github.com/AfroSpartan) All the stuff how to use signals with c# be it with connect or += and -= are nicely written in the docs: https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_signals.html#c-signals linked earlier. You can use c# events, godot signals, combine it in your code.

Yeah, I fixed it with adding an `_ExitTree()` deconstructor. My question was regarding the fact that it doesn't seem that C# is integrated in a way where Godot's disposal is propagated up into the C# layer in this case, causing C# code to behave in a way you wouldn't expect.
